examples/cpu/ccl_example.cpp

```diff
@@ -0,0 +1,136 @@
+        0, 1, 2, 3, 4, 6, 8, 11, 16, 23, 32, 45, 64, 91, 128, 181, 256, 362,
+        512, 724, 1024, 1448, 2048, 2896, 4096, 5793, 8192
+    };
+
+    static std::size_t max_width = 50;
+
+    std::vector<std::size_t> bins(bins_edges.size());
+
+    for (const traccc::cell_collection & module : modules) {
+        std::size_t count = module.items.size();
+
+        for (std::size_t i = 0; i < bins_edges.size(); ++i) {
+            if (
+                count >= bins_edges[i] &&
+                (i + 1 >= bins_edges.size() || count < bins_edges[i + 1])
+            ) {
+                ++bins[i];
+                break;
+            }
+        }
+    }
+
+    std::size_t max = *std::max_element(bins.begin(), bins.end());
+
+    std::size_t per_pixel = std::max(
+        static_cast<std::size_t>(1),
+        (max + (max % max_width)) / max_width
+    );
+
+    std::cout << "\nNon-zero pixels per module" << std::endl;
+
+    std::cout << std::setw(5) << "Min" << "   " << std::setw(5) << "Max" <<
+    " | " << std::setw(5) << "Count" << std::endl;
+
+    for (std::size_t i = 0; i < bins.size(); ++i) {
+        std::cout << std::setw(5) << bins_edges[i] << " - " << std::setw(5) <<
+        (i + 1 >= bins.size() ? "inf" : std::to_string(bins_edges[i + 1] - 1)) <<
+        " | " << std::setw(5) << bins[i] << " ";
+
+        for (std::size_t j = 0; j < bins[i]; j += per_pixel) {
+            std::cout << "*";
+        }
+
+        std::cout << std::endl;
+    }
+}
+
+int main(
+    int argc,
+    char * argv[]
+) {
+    if (argc < 2){
+        std::cout << "Not enough arguments, minimum requirement: " << std::endl;
+        std::cout << argv[0] << " <event_file>" << std::endl;
+        return -1;
+    }
+
+    std::string event_file = std::string(argv[1]);
+
+    std::cout << "Running " << argv[0] << " on " << event_file << std::endl;
+
+    traccc::component_connection cc;
+
+    auto time_read_start = std::chrono::high_resolution_clock::now();
+
+    traccc::cell_reader creader(event_file, {"geometry_id", "hit_id", "cannel0", "channel1", "activation", "time"});
+    traccc::cell_container modules = traccc::read_cells(creader);
+
+    auto time_read_end = std::chrono::high_resolution_clock::now();
+
+    print_statistics(modules);
+
+    auto time_process_p1 = std::chrono::high_resolution_clock::now();
+
+    for (const traccc::cell_collection & module : modules) {
+        traccc::cluster_collection clusters_per_module = cc(module);
+    }
+
+    auto time_process_p2 = std::chrono::high_resolution_clock::now();
+
+    for (std::size_t i = 0; i < 10; ++i) {
+        for (const traccc::cell_collection & module : modules) {
+            traccc::cluster_collection clusters_per_module = cc(module);
+        }
+    }
+
+    auto time_process_p3 = std::chrono::high_resolution_clock::now();
+
+    for (const traccc::cell_collection & module : modules) {
+        traccc::cluster_collection clusters_per_module = cc(module);
```

Comment:
:thinking: I wonder if the code should explicitly do something with the output variable to make sure that the code is getting executed. As a very aggressive compiler optimisation may decide not to execute this line at all, since its output is not needed anywhere later on.

Doing something as silly as

```c++
std::size_t dummy = 0;
for (const traccc::cell_collection & module : modules) {
       dummy += cc(module).size();
}
...
std::cout << ( dummy % 2 ? ' ' : ' ' );
```

should force the compiler to not be able to "optimise out" this code. (Of course we would adjust this once/if the cluster EDM is updated...)

Reply:
This optimization barrier is too simple, the compiler can optimize it out into `std::cout << ' ';` and then proceed to remove the computation of `dummy`.

Here are the compiler optimization barriers that I typically use. These will work on all compilers that support the GNU dialect of inline assembler (i.e. anything but MSVC, which we probably don't care about here). You can do everything with the `assume_accessed_global` barrier, but the `_local` barriers may lead to nicer codegen when applied judiciously.

```cpp
  // Forces compiler to assume that any data directly or indirectly targeted by
  // the reference has been read and written. Easy to use, but may affect compiler
  // memory access optimizations for data behind unrelated pointers...
  template <typename T>
  void assume_accessed_global(const T& data) {
    asm("" : : "g"(&data) : "memory");
  }

  // Forces compiler to assume that a configurable number of elements behind the
  // target pointer have been read and written, will not account for transitive
  // pointers but will not affect compiler optimizations for other memory regions.
  template <typename T, size_t ACCESSED_ELEMS=1>
  void assume_accessed_local(const T* ptr) {
    const auto array_ptr = reinterpret_cast<T(*)[ACCESSED_ELEMS]>(const_cast<T*>(ptr));
    asm("" : "+m"(*array_ptr) : : );
  }

  // Forces compiler to assume that a configurable number of elements behind the
  // target pointer have been read, will not account for transitive pointers but
  // will not affect compiler optimizations for other memory regions.
  template <typename T, size_t ACCESSED_ELEMS=1>
  void assume_read_local(const T* ptr) {
    const auto array_ptr = reinterpret_cast<const T(*)[ACCESSED_ELEMS]>(ptr);
    asm("" : : "m"(*array_ptr) : );
  }

  // Forces compiler to assume that a configurable number of elements behind the
  // target pointer have been written, will not account for transitive pointers
  // but will not affect compiler optimizations for other memory regions.
  template <typename T, size_t ACCESSED_ELEMS=1>
  void assume_written_local(const T* ptr) {
    const auto array_ptr = reinterpret_cast<T(*)[ACCESSED_ELEMS]>(const_cast<T*>(ptr));
    asm("" : "=m"(*array_ptr) : : );
  }
  ```

Reply:
I'd say for now it's probably fine as it is. A compiler would have to have _very_ good static analysis capabilities to decide to optimize out this function call, and my gcc is leaving it as it. This program is also not really meant to be a benchmark, but rather an example of how the CCL code might be used. We can always look back at this later, if it turns out to be an issue. :slightly_smiling_face:

Reply:
I'm fine with that. :+1:

For the "final" benchmarks in the project we should come up with some clever code for this, but in this particular executable it's not a priority. Not just yet.

Hadrien, I would prefer if we could find some clever way of doing this without relying on platform specific tricks. I think that would be possible as well. (Maybe not in a completely generic way, but at least in a "return value specific" one.) But let's table that discussion until we really need it...

P.S. Let's see if the thumbs-up emoji has an effect on this PR's approval status... :thinking:

Reply:
@krasznaa Note that these tricks are only compiler specific, not CPU specific (because the ASM block body is empty, only the input/output/clobbers are used as an optimization barrier). But yeah, portable optimization barriers are hard.

Without using compiler-specific extensions, the best trick I know is to escape a pointer to a separate compilation unit and write in big red bold letters in your documentation WHATEVER YOU DO, DO NOT ENABLE LTO, and then someday somebody needs LTO, and the overhead of that trick is much higher than that of inline assembly trick. So given that pretty much everyone but MSVC supports the GNU C inline assembly syntax well enough for my needs, I usually don't bother.